sdk/lib/rtl/heap.c

```diff
@@ -1087,8 +1100,19 @@ RtlpInitializeHeapSegment(IN OUT PHEAP Heap,
                     FreeEntry = Segment->FirstEntry;
                     FreeEntry->PreviousSize = Segment->Entry.Size;
                     FreeEntry->SegmentOffset = SegmentIndex;
-                    FreeEntry->Size = GuardEntry->PreviousSize;
+                    FreeEntry->Size = PreviousSize;
                     FreeEntry->Flags = 0;
+
+                    /* Register the Free Heap Entry */
+                    FreeEntry = (PHEAP_ENTRY)RtlpInsertFreeBlock(Heap, (PHEAP_FREE_ENTRY)FreeEntry, PreviousSize);
+
+                    /* Fill guard entry */
+                    ASSERT(FreeEntry->Size != 0);
+                    ASSERT(FreeEntry != GuardEntry);
+                    ASSERT(FreeEntry->Size != 1); /* FIXME: we need to double the guard entry */
```

Comment:
I don't get the FIXME comment. The guard entry needs to be doubled in case the free entry before it is too small to be suitable for an alloc, which is size == 1. The assert is exactly the contrary of that.

Reply:
The assert is written such that it fires if size == 1, because that's the unimplemented code path.
If the general approach is acceptable, this would turn into `if (FreeEntry->Size == 1) { /* double the guard entry */ }`

Reply:
In that case, please use UNIMPLEMENTED_DBGBREAK
ReactOS code is crippled with asserts where we're not sure if those are asserts or unimplemented code paths.
But in that case, I do think this is a valid assert, see in RtlpInsertFreeBlock:
```
            /* Special compensation if it goes above limit just by 1 */
            if (BlockSize == (HEAP_MAX_BLOCK_SIZE + 1))
                Size -= 16;
```